core/optimization/src/main/java/it/unibz/inf/ontop/iq/optimizer/impl/PreventDistinctOptimizerImpl.java

```diff
@@ -0,0 +1,225 @@
+import it.unibz.inf.ontop.model.term.*;
+import it.unibz.inf.ontop.model.type.DBTermType;
+import it.unibz.inf.ontop.model.type.TermTypeInference;
+import it.unibz.inf.ontop.substitution.SubstitutionFactory;
+import it.unibz.inf.ontop.utils.ImmutableCollectors;
+import it.unibz.inf.ontop.utils.VariableGenerator;
+
+import javax.inject.Inject;
+import java.util.Optional;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+public class PreventDistinctOptimizerImpl implements PreventDistinctOptimizer {
+
+    private final IntermediateQueryFactory iqFactory;
+    private final OptimizationSingletons optimizationSingletons;
+
+    @Inject
+    private PreventDistinctOptimizerImpl(IntermediateQueryFactory iqFactory, OptimizationSingletons optimizationSingletons) {
+        this.iqFactory = iqFactory;
+        this.optimizationSingletons = optimizationSingletons;
+    }
+
+    @Override
+    public IQ optimize(IQ query) {
+        VariableGenerator variableGenerator = query.getVariableGenerator();
+        IQTreeTransformer transformer = createTransformer(variableGenerator);
+        IQTree newTree = transformer.transform(query.getTree());
+        return iqFactory.createIQ(query.getProjectionAtom(), newTree);
+    }
+
+    protected IQTreeTransformer createTransformer(VariableGenerator variableGenerator) {
+        return new PreventDistinctTransformer(variableGenerator, optimizationSingletons);
+    }
+
+    private static class PreventDistinctTransformer extends DefaultRecursiveIQTreeVisitingTransformer {
+
+        private final TermFactory termFactory;
+        private final SubstitutionFactory substitutionFactory;
+        private final VariableGenerator variableGenerator;
+
+        public PreventDistinctTransformer(VariableGenerator variableGenerator, OptimizationSingletons optimizationSingletons) {
+            super(optimizationSingletons.getCoreSingletons());
+            this.variableGenerator = variableGenerator;
+            this.termFactory = optimizationSingletons.getCoreSingletons().getTermFactory();
+            this.substitutionFactory = optimizationSingletons.getCoreSingletons().getSubstitutionFactory();
+        }
+
+        @Override
+        public IQTree transformConstruction(IQTree tree, ConstructionNode rootNode, IQTree child) {
+            var decomposition = Decomposition.decomposeTree(child);
+            if(decomposition.distinctNode.isPresent()) {
+                return performTransformation(rootNode, tree, decomposition);
+            }
+            return super.transformConstruction(tree, rootNode, child);
+        }
+
+        private IQTree performTransformation(ConstructionNode constructionNode, IQTree originalTree, Decomposition decomposition) {
+            var subtree = decomposition.descendantTree;
+
+
+            //Find terms that need to be pushed below the distinct.
+            ImmutableMap.Builder<Variable, ImmutableTerm> substitutionBuilder = ImmutableMap.builder();
```

Comment:
Please find a better name for it. It is not building a substitution object but a map. From its name, it is a bit hard to understand what this map is about.